### ISSUE TITLE
Makefile is broken when combining "binary" / "dynbinary" with other targets

### ISSUE THREAD

**Issue by thaJeztah** [moby/moby#40088]
introduced by https://github.com/moby/moby/pull/39340

TL;DR; I know what's causing this, but will have to look how to best solve; thinking of having separate `build-image` and `build-binary` targets (one of the could be named just `build`); 

- `build-binary` would output a binary (run build with `--output..`)
- `build-image` would build the dev image

The problem introduced by https://github.com/moby/moby/pull/39340 is that after that change, the `binary`, `dynbinary`, and `cross` targets append a `--output` to `DOCKER_BUILD_ARGS`; https://github.com/moby/moby/blob/d56adcf0ece542a30fad0e54ea9c1ca3b7e6915c/Makefile#L151-L159

While this works when running those targets _separately_ (e.g. just `make binary`), it doesn't work when combining them with other targets, as it would

- when combined with (e.g.) the `shell` or `test-integration` target; only build the Dockerfile with `--output`, so no image will be built (only a binary)
- any combination of `binary`, `dynbinary`, and/or `cross` would append _multiple_ `--output` and multiple `--target` options to `DOCKER_BUILD_ARGS`, which currently isn't supported

For example, combining `binary` and `shell`;

```bash
docker build  --output=bundles/ --target=binary --build-arg=CROSS=false  -t "docker-dev:support-xattrs" -f "Dockerfile" .

[+] Building 19.2s (73/73) FINISHED 
....
=> exporting to client
 => => copying files 195.44MB
...
docker run --rm -i --privileged  .......  "docker-dev:support-xattrs" bash
Unable to find image 'docker-dev:support-xattrs' locally
docker: Error response from daemon: manifest for docker-dev:support-xattrs not found: manifest unknown: manifest unknown.
```

And combining `binary` and `dynbinary`;

```console
make binary dynbinary
docker build  --output=bundles/ --target=binary --build-arg=CROSS=false  -t "docker-dev:support-xattrs" -f "Dockerfile" .
[+] Building 9.6s (73/73) FINISHED
...
 => exporting to client
 => => copying files 195.48MB
make: Nothing to be done for `dynbinary'.
```

Here, `dynbinary` became a no-op, because it effectively is an alias for `build` (but modifying `DOCKER_BUILD_ARGS` first), however `make` detects that the `build` target has already been executed, thus skips this.

## Workaround

instead of combining, run the make targets separately;

```console
make binary
docker build  --output=bundles/ --target=binary --build-arg=CROSS=false  -t "docker-dev:support-xattrs" -f "Dockerfile" .
[+] Building 9.6s (73/73) FINISHED
...
 => exporting to client
 => => copying files 195.48MB

make shell
docker build  --target=final --build-arg=CROSS=false  -t "docker-dev:support-xattrs" -f "Dockerfile" .
[+] Building 2.2s (71/71) FINISHED
...
 => => exporting layers
 => => writing image sha256:cb927fd7a4e4af4de5200bad984a66fc7e7d32901447889231af6413dee801bc
 => => naming to docker.io/library/docker-dev:support-xattrs
docker run --rm -i --privileged  .......  "docker-dev:support-xattrs" bash
...
```

/cc @cpuguy83

**Comment by cpuguy83** [comment]
We should probably call `docker build` directly in each of these binary steps.

**Comment by thaJeztah** [comment]
Yes, I think that'd work.

Also thinking of adding a `BUILD` or `DOCKER_BUILD` variable, which would be either set to `docker build` or `buildx build`, depending on the situation.

**Comment by cpuguy83** [comment]
Yeah, working on just that :)

**Comment by thaJeztah** [comment]
👍 thanks! Our makefile is way too complicated, LOL

**Comment by kolyshkin** [comment]
Found that this behavior is [documented in GNU make](https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html):

> Be aware that a given prerequisite will only be built once per invocation of make, at most. If the same file is a prerequisite of multiple targets, and each of those targets has a different value for the same target-specific variable, then the first target to be built will cause that prerequisite to be built and the prerequisite will inherit the target-specific value from the first target. It will ignore the target-specific values from any other targets.

... essentially meaning that `build` target will only be built once.

Seems like the solution, as mentioned in comments above, is to call build directly (using as much var assignments as possible to dedup the whole thing)
